File: 0206_Reverse_Linked_List/solution.go
// https://leetcode.com/problems/reverse-linked-list
package leetcode0206
import (
	. "../common"
)
/*
	
 */

func reverseList(head *ListNode) *ListNode {
		// prev 是所有已经逆转的节点的head
		var prev *ListNode

		// head 是下一个被逆转的节点
		for head != nil {
			// 让next指向head.Next, 免得head.Next不见了.
			next := head.Next
			// head称为已经逆转的节点的新head
			head.Next = prev
			// 让prev重新称为所有已被逆转节点的head
			prev = head
			// 让head指向下一个未被逆转的节点
			head = next
		}

		return prev
}

File: 0206_Reverse_Linked_List/solution.cpp
/*
206. Reverse Linked List

Level: Easy

https://leetcode.com/problems/reverse-linked-list
*/
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>

#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
  ListNode *reverseList(ListNode *head)
  {
    ListNode *prev = nullptr;
    ListNode *curr = head;

    while (curr)
    {
      ListNode *next = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next;
    }
    return prev;
  }
};